### DISCUSSION TITLE
How should I handle the Error "Unique constraint failed on the constraint: `PRIMARY`"

### DISCUSSION THREAD

**Discussion by joisun** [prisma/prisma#16709] [Q&A] [answered]
How should I handle the Error "Unique constraint failed on the constraint: `PRIMARY`"

I'm try to build a blog system use prisma, and when I try to upsert An article/post , so if the `artid` (which is article id, the PRIMARY KEY of article table) exists in the data from frontend web page editor, I need to update the article data, otherwise I need create a new one. 
seems there is no logical error, but I failed.  can anyone help me out with this ? thanks!

(btw, here is my prisma file if needed [link](https://github.com/jaycethanks/zhongyi-blog-server/blob/master/prisma/schema.prisma))

Here is the code: 

```ts
 async upsert(userid: string, createAdminArticleDto: CreateAdminArticleDto) {
    const {
      artid,
      title,
      content,
      isbanner,
      description,
      cover,
      password,
      visible,
      tags,
      status,
      column,
      category,
    } = createAdminArticleDto;
    try {
      let res = null;
      const data = {
        title,
        content,
        author: {
          connect: {
            userid: userid,
          },
        },
        banner: isbanner ? 1 : 0,
        description,
        cover,
        password,
        visible,
        status,
        column: column
          ? {
              connect: {
                colid: column ? column : undefined,
              },
            }
          : undefined,
        category: category
          ? {
              connect: {
                catid: category,
              },
            }
          : undefined,
        tags:
          tags.length > 0
            ? {
                create: tags.map((_tag) => ({
                  tag: {
                    connectOrCreate: {
                      create: {
                        name: _tag,
                        userid: userid,
                      },
                      where: {
                        tagid: _tag,
                      },
                    },
                  },
                })),
              }
            : undefined,
      };
      if (artid) {
        // update
        res = await this.prisma.article.update({
          where: {
            artid,
          },
          data,
        });
      } else {
        // create
        res = await this.prisma.article.create({
          data,
        });
      }
      return Result.okData(res);
    } catch (error) {
      console.log('[error]: ', error);
      return Result.error('文章添加出现了预料之外的错误!', 5002);
    }
  }

  ```

  I also tried use the function `upsert()` directly, but met the seem error.

**Comment by nurul3101** [comment]
Hi @jaycethanks 👋 
Is this the full error message that you get?
Do you get any information on which column is the unique constraint failing?
From the schema, I can see that tag name has a unique constraint, are all the names in tags array unique?

**Reply by joisun** [reply]
yes, all tag's name is unique, and finnally I found the error caused by the relation fields' id, when I set some fields are required to undefined, so I changed theme and It works fine. thanks:

**Comment by joisun** [comment, marked as answer]
This problem I met caused by I trying to set some Article's relation fields that are required to be undefined.
Just set them to be optional seems gonna work, but there's a  "logical paradox", cause, the field `category` is designed to be required. 
and here are the changes I made in my case, hope it will help someone encountered the same error.

```diff
model Article {
  artid    String @id @default(uuid())
  sort     Int    @default(0)
  author   User   @relation(fields: [authorId], references: [userid])
  authorId String @map("author_id")

  title       String
  description String?
  content     String  @db.LongText

  keywords String @default("")
  status   Int    @default(0) //0: 草稿 1:发布 3：删除

  cover    String    @default("")
  visible  Int       @default(1) // 0: 不可见 1:可见
  password String? // 加密文章
  comments Comment[]

  allowComment Int       @default(1) @map("allow_comment") // 1: 允许评论 0: 禁止评论
  liking       Int       @default(0)
  readers      Int       @default(0)
  banner       Int       @default(0) //0: 不在 banner 区域展示， 1 banner 区域的顺序
- category     Category  @relation(fields: [catid], references: [catid]) // 分类关联
+ category     Category? @relation(fields: [catid], references: [catid]) // 分类关联
- catid        String
+ catid        String?
- column       Column    @relation(fields: [colid], references: [colid]) // 专栏关联
+ column       Column?   @relation(fields: [colid], references: [colid]) // 专栏关联
- colid        String
+ colid        String?

  tags TagsOnArticles[] // 标签关联

  createdAt DateTime @default(now()) @map("created_at")
  updatedAt DateTime @default(now()) @updatedAt @map("updated_at")

  @@map("t_articles")
}
```

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
